### DISCUSSION TITLE
Missing Entitlements on App Store Connect uploaded Builds

### DISCUSSION THREAD

**Discussion by nevotheless** [codemagic-ci-cd/codemagic-docs#2173] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1.  Edit the Info.plist through the Xcode Capabilities Editor as documented.
2. Have Codemagic handle `export_options.plist` generation 
3. Win... not so much in this case ☹️ 

I don't quite know what file is ultimately is responsible for the correct capabilities. 

But what i've done is

- Check the provisioning profiles in Apple Developer Portal and i can definitely verify that the codemagic auto-generated profiles have the correct capabilities (Push Notifications and Associated Domains)
- Removed and Re-added Capabilities through Xcode, this changes `Info.plist` and `RunnerRelease.entitlements` Files but no changes visibile through `git diff`.

I also checked the Entitlement relevant Build Configuration Links as seen here:
<img width="642" alt="image" src="https://github.com/codemagic-ci-cd/codemagic-docs/assets/779797/e5969f78-4432-440c-a262-1ab8247f6153">
These files are seemingly getting ignored.

I suspect the generated `export_options.plist` during the iOS Build step is getting stitched together wrong maybe? But example files don't have anything related to capabilities so this might also be absolutely wrong and i just have no other idea anymore what could be wrong here.

### Expected results

Having the mentioned Capabilities listed in App Store Connect and not getting a Email about the Infamous `ITMS-90078`.

### Actual results

<img width="462" alt="image" src="https://github.com/codemagic-ci-cd/codemagic-docs/assets/779797/c270302e-6fed-49fc-9b27-8a740db2d540">

As well as the `ITMS-90078` Email Notification about missing Entitlement.

### Build id (optional)

6544e75df65ed22c8e0bf30a

**Comment by nevotheless** [comment]
Further investigation hints towards the `codesign` command. Which doesn't seem to sign the binaries with the appropriate entitlement file?

**Comment by yusufceyhun** [comment]
I guess ITMS-90078 is about push notifications. If you want to enable push notification, you need to configure XCode correctly.

**You need to have 2 capabilities added to your XCode:
1 - Push Notifications
2 - Background Modes 
Ênsure that both the "Background fetch" and the "Remote notifications" sub-modes are enabled.

https://firebase.flutter.dev/docs/messaging/apple-integration/ this will help for the configurations.**

I don't recommend you to manually edit files but i think you need to manually edit AppDelegate.swift

```
import UIKit
import Flutter

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {    
    if #available(iOS 10.0, *) {
      UNUserNotificationCenter.current().delegate = self as? UNUserNotificationCenterDelegate
    }
    
    GeneratedPluginRegistrant.register(with: self)

    application.registerForRemoteNotifications()

    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
}

```
it suppose to look like this. Inside your Runner.entitlements don't forget to enable production:

```
<key>aps-environment</key>
<string>production</string>
```

I always debug with release mode on maybe this would help either.

**Reply by nevotheless** [reply]
We followed the Flutter docs for this and enabled everything accordingly inside xcode already.
Besides Push Entitlement we are also trying to add **associated domains** entitlement which is also not existent when i try to check the artifacts signature.

So this issue is not limited to push notifications.
 
```bash
» codesign -d --entitlements - Payload/Runner.app
Executable=/Users/asdf/Downloads/downloaded_artifact_and_extracted_ipa_as_zip/Payload/Runner.app/Runner
[Dict]
	[Key] application-identifier
	[Value]
		[String] XXXXXXXX.de.somecompany.app
	[Key] beta-reports-active
	[Value]
		[Bool] true
	[Key] com.apple.developer.team-identifier
	[Value]
		[String] XXXXXXXX
	[Key] get-task-allow
	[Value]
		[Bool] false

**Reply by nevotheless** [reply]
We can workaround re-signing the binary with the appropriate `ios/Runner/RunnerRelease.entitlements` File but this feels very wrong and is not possible in codemagic with the workflow builder. 

Something is preventing the entitlements getting attached to the signature :(

**Comment by nevotheless** [comment, marked as answer]
I found out that out project files were corrupted in some way. 😞 Now its fixed. (Had nothing to do with codemagic)
